### PR TITLE
VESA video mode - 800x600x16bit

### PR DESCRIPTION
**Pull request by jabedude**
Hello all, this PR checks if VESA 2.0 or 3.0 is supported and then enables it with fixed 800x600. This is very rough, so please let me know what else we need to do.

TODO:

- [x] Proper color handling
- [ ] Query the framebuffer address from hardware instead of hardcoding it (if possible)
- [ ] Pass the framebuffer address in BootInfo

### CODE REVIEW

**Comment by phil-opp** [comment]
Thanks a lot for tackling this!

We don't want to hard code the video mode for the user, but instead make it configurable through cargo features. We currently support [vga text mode and a simple vga graphics mode (320x200)](https://github.com/rust-osdev/bootloader/blob/4f87e5a2efc239ca301fcf470781a0ff66ce268b/src/main.rs#L34-L37). Could you place the VESA code in a new file in `src/video_mode` and add a cargo feature for enabling it, like it is done for the other video modes?

**Comment by jabedude** [comment]
Sure thing!

**Comment by phil-opp** [comment]
Thanks, looks good now!

Were you able to print something to the screen through this? I just tried to update the `src/printer/mod.rs` module (used for printing panic messages) with a new vesa printer, but I didn't know the framebuffer location.

**Comment by jabedude** [comment]
I meant to ping you, I actually was hoping to get some help with testing printing with the new video mode. I should be able to get you the framebuffer address.

**Comment by phil-opp** [comment]
Sure!

Printing should work similar to the [vga_320x200](https://github.com/rust-osdev/bootloader/blob/master/src/printer/vga_320x200.rs) module, we only need the framebuffer address and the pixel format (i.e. how many bytes per pixel, how are the RGB channels encoded).

**Comment by jabedude** [comment]
I found the framebuffer address for qemu: 0xFD000000. The pixel format is 5 red bits, 6 green, and 5  blue bits. I'm still having issues trying to modify vga_320x200 to print correctly

**Comment by phil-opp** [comment]
@jabedude Thanks! The problem was that we need to add a page table mapping for the physical address `0xFD000000` to the page table. I pushed a commit (I hope that is fine with you) that maps the framebuffer to the 500th entry of the level 2 page table. It also adds a new printer, which is the vga_320x200 printer with some modifications. Now it looks like this:

image

I didn't add any code for color handling, so the colors are a bit strange. I deliberately didn't paint the screen back in `clear_screen` so that we see that really the whole screen is blanked.

**Comment by ethindp** [comment]
Here's what I'm wondering: will this bootloader be modified sometime
to make it possible to boot on real hardware? It seems like the
bootloader goes off virtualized/emulated addresses, at least for
anything that's not 0xb8000. I would love to be able to boot my kernel
on real hardware with this thing!
If not, how can I get it to do that?

On 7/6/19, Philipp Oppermann <notifications@github.com> wrote:
> @jabedude Thanks! The problem was that we need to add a page table mapping
> for the physical address `0xFD000000` to the page table. I pushed a commit
> (I hope that is fine with you) that maps the framebuffer to the 500th entry
> of the level 2 page table. It also adds a new printer, which is the
> vga_320x200 printer with some modifications. Now it looks like this:
>
> image
>
> I didn't add any code for color handling, so the colors are a bit strange. I
> deliberately didn't paint the screen back in `clear_screen` so that we see
> that really the whole screen is blanked.
>
> --
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/rust-osdev/bootloader/pull/60#issuecomment-508908451


-- 
Signed,
Ethin D. Probst

**Comment by phil-opp** [comment]
@ethindp For me the bootloader already works on real hardware. What problems do you have exactly?

Also, could you elaborate your statement about "virtualized/emulated addresses"? Since you replied to this pull request, I assume that you mean using the address `0xFD000000`? This is just a stop-gap solution for experimentation, the plan is to query this address from the VESA later.

**Comment by ethindp** [comment]
Aha, I see. Well, I was following yorublog_os tutorial and wanted to
make an ISO with it. I might've passed the wrong args to mkisofs, but
I created the IsO and the bootloader got stuck on stage 1 (I think
that there is an issue for this).

On 7/6/19, Philipp Oppermann <notifications@github.com> wrote:
> @ethindp For me the bootloader already works on real hardware. What problems
> do you have exactly?
>
> Also, could you elaborate your statement about "virtualized/emulated
> addresses"? Since you replied to this pull request, I assume that you mean
> using the address `0xFD000000`? This is just a stop-gap solution for
> experimentation, the plan is to query this address from the VESA later.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/rust-osdev/bootloader/pull/60#issuecomment-508941217


-- 
Signed,
Ethin D. Probst

**Comment by phil-opp** [comment]
Have you tried the instructions at https://os.phil-opp.com/minimal-rust-kernel/#real-machine? The `bootimage` tool already creates a hard disk image, so you don't need `mkisofs` for this (I'm not sure about CD images though).

As a side note: Can we continue the conversation in a new issue? This issue issue is about adding support for the VESA video mode, so I think a separate issue would fit better.

**Comment by jabedude** [comment]
@phil-opp That is awesome! Of course, I was hoping to collaborate in this PR (I have no prior experience with VESA/BIOS printing).

**Comment by phil-opp** [comment]
@jabedude Great! I updated the PR description with a todo list. Do you think anything else is missing?

**Comment by jabedude** [comment]
The TODOs look good to me! I pushed a commit to query the framebuffer address, but I'm not sure the best way to test that.

Related to the third TODO, how can I get access to the `VESAfb` variable from `load_elf()`?

**Comment by phil-opp** [comment]
I think we can do the same as in https://github.com/rust-osdev/bootloader/blob/b5d43caa72173b502dd1eabcee29d534ef24741a/src/main.rs#L66-L77 to get access to `VESAfb` and then pass as argument to `load_elf`. To use it in the printer, we need to initialize it at runtime with the given framebuffer address, e.g. through an `init` method.

File: src/printer/vesa_800x600.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                    for (x, bit) in (0..8).enumerate() {
+                        if *byte & (1 << bit) == 0 {
+                            continue;
+                        }
+                        let color = 0xff;
+                        vga_buffer[(y_pos + y) * SCREEN_WIDTH + x_pos + x] = color;
+                    }
+                }
+            }
+            _ => panic!("unprintable character"),
```
Comment (bjorn3) [inline src/printer/vesa_800x600.rs]: Please dont panic in this case. Maybe print `?` or something like that.

Comment (adeadman) [reply src/printer/vesa_800x600.rs]: The panic won't work as the writer is already locked for the write. One way around this deadlock situation is to separate the screen drawing from the character lookup in the font map - or to validate that all characters in the string are in the character set supported before calling into the static object.

**Comment by adeadman** [comment]
> I think we can do the same as in
> 
> https://github.com/rust-osdev/bootloader/blob/b5d43caa72173b502dd1eabcee29d534ef24741a/src/main.rs#L66-L77
> 
> to get access to `VESAfb` and then pass as argument to `load_elf`. To use it in the printer, we need to initialize it at runtime with the given framebuffer address, e.g. through an `init` method.

Thinking about this, for flexibility we probably want to load an entire VBEModeInfo struct into the kernel space so we have flexibility for choosing different modes in the bootloader. 

The framebuffer mapping to virtual memory currently hardcodes the framebuffer address for qemu and the size for an 800x600x16 mode, so getting it to work with an arbitrary framebuffer address (e.g. from real hardware) using the vbe mode info is a bit harder. I've been working on it on and off for the last couple weeks and am wondering if it would be easier or even possible to do the memory mapping from the rust code, or if it has to be done while in protected mode stage 3 before we jump to the kernel loading.

**Comment by ethindp** [comment]
Hate to go OT, but would it be possible for me to boot my kernel using
Grub? Or would that not work (i.e. if I were to use grub-mkrescue)?

On 8/10/19, Aaron Deadman <notifications@github.com> wrote:
>> I think we can do the same as in
>>
>> https://github.com/rust-osdev/bootloader/blob/b5d43caa72173b502dd1eabcee29d534ef24741a/src/main.rs#L66-L77
>>
>> to get access to `VESAfb` and then pass as argument to `load_elf`. To use
>> it in the printer, we need to initialize it at runtime with the given
>> framebuffer address, e.g. through an `init` method.
>
> Thinking about this, for flexibility we probably want to load an entire
> VBEModeInfo struct into the kernel space so we have flexibility for choosing
> different modes in the bootloader.
>
> The framebuffer mapping to virtual memory currently hardcodes the
> framebuffer address for qemu and the size for an 800x600x16 mode, so getting
> it to work with an arbitrary framebuffer address (e.g. from real hardware)
> using the vbe mode info is a bit harder. I've been working on it on and off
> for the last couple weeks and am wondering if it would be easier or even
> possible to do the memory mapping from the rust code, or if it has to be
> done while in protected mode stage 3 before we jump to the kernel loading.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/rust-osdev/bootloader/pull/60#issuecomment-520190457


-- 
Signed,
Ethin D. Probst

**Comment by phil-opp** [comment]
@ethindp See <https://github.com/rust-osdev/bootloader/blob/master/doc/chainloading.md> and https://github.com/rust-osdev/bootloader/issues/49. In the future, please just open a new issue instead of commenting on an unrelated issue.

**Comment by phil-opp** [comment]
@adeadman 

>  I've been working on it on and off for the last couple weeks and am wondering if it would be easier or even possible to do the memory mapping from the rust code, or if it has to be done while in protected mode stage 3 before we jump to the kernel loading.

If we don't access it before the mapping, we can probably map it later. However, we need to adjust our panic printer for that (i.e. it shouldn't try to print to the VESA framebuffer before it is mapped).

**Comment by phil-opp** [comment]
Given that the bootloader was changed significantly since this PR (e.g. in #130), I'm going to give this a close. We still don't support choosing custom framebuffer resolutions though, so if anyone wants to tackle this, please let me know!

Thanks you @jabedude for all your work!

**Pull request closed without merging**